Improve ProcessGroup RpcBackendOptions Constructor API

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #34139 Apply clang-format to RPC files
* #34095 Use double quotes in C++ to stay consistent with Python RPC docs
* **#34081 Improve ProcessGroup RpcBackendOptions Constructor API**

Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

Differential Revision: [D20227344](https://our.internmc.facebook.com/intern/diff/D20227344)

mrshenli committed:
Improve ProcessGroup RpcBackendOptions Constructor API

Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

[ghstack-poisoned]

mrshenli commented:
![Screen Shot 2020-03-02 at 4 21 14 PM](https://user-images.githubusercontent.com/16999635/75722021-614dbd80-5ca7-11ea-97e6-f2062dcc9e08.png)

mrshenli commented:
Lint failure is on the line I didn't touch:

```
    path: 'torch/nn/utils/rnn.py',
    start_line: 153,
    end_line: 153,
    start_column: 136,
    end_column: 136,
    annotation_level: 'failure',
    message: '[B950] line too long (135 > 120 characters)'
  }
```

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

[ghstack-poisoned]

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

[ghstack-poisoned]

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

[ghstack-poisoned]

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

[ghstack-poisoned]

rohan-varma commented on torch/distributed/rpc/constants.py:
```diff
@@ -1,12 +1,16 @@
-DEFAULT_RPC_TIMEOUT = timedelta(seconds=60)
-DEFAULT_INIT_METHOD = "env://"
+from . import (
+    _DEFAULT_RPC_TIMEOUT,
```
Is this change because they are defined in C++ and exposed to python now?

mrshenli replied:
Yes. We need a default value for the `ProcessGroupRpcBackendOptions` constructor arguments, so that these defaults need to live in C++. Important them here to avoid duplications.

rohan-varma commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1542,6 +1542,26 @@ def test_get_rpc_timeout(self):
+        )
+
+        info = rpc.api._get_current_rpc_agent().get_debug_info()
+        self.assertEqual(int(info["agent.num_idle_threads"]), NUM_THREADS)
```
Are we guaranteed that there will be no work that takes up a thread so that this test won't be flaky?

mrshenli replied:
good point! I should have used `thread_pool_size`.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -12,9 +12,17 @@ namespace torch {
+      std::chrono::milliseconds rpc_timeout,
+      std::string init_method)
+      : RpcBackendOptions(rpc_timeout, init_method),
+        numSendRecvThreads_(num_send_recv_threads) {}
```
Would we want to do some error checking here (or higher in Python layer) and throw if an invalid number such as negative is passed in?

rohan-varma commented on docs/source/rpc.rst:
```diff
@@ -103,7 +103,9 @@ to appropriate devices if necessary.
 .. autoclass:: RpcBackendOptions
     :members:
-
+.. autoclass:: ProcessGroupRpcBackendOptions
```
Thanks for updating the docs!

rohan-varma commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -39,21 +39,25 @@ PyObject* rpc_init(PyObject* /* unused */) {
-            An instance of this class can be passed in to :meth:`~torch.distributed.rpc.init_rpc`
-            in order to initialize RPC with specific configurations, such as the
-             RPC timeout and init_method to be used. )")
+          R"(An abstract structure encapsulating the options passed into the RPC
```
We can't directly pass in `RpcBackendOptions`, correct? Should we explicitly state this, and maybe remove `RpcBackendOptions` from the public docs? It doesn't seem like the user will ever have to construct this - they'd always use the process group variant.

mrshenli replied:
Right, we cannot pass `RpcBackendOptions` directly, let me remove it then.

rohan-varma reviewed (approved):
Thanks for adding this! Just had a couple minor comments after which we can land.

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

Differential Revision: [D20227344](https://our.internmc.facebook.com/intern/diff/D20227344)

[ghstack-poisoned]

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

Differential Revision: [D20227344](https://our.internmc.facebook.com/intern/diff/D20227344)

[ghstack-poisoned]

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

Differential Revision: [D20227344](https://our.internmc.facebook.com/intern/diff/D20227344)

[ghstack-poisoned]

mrshenli committed:
Update on "Improve ProcessGroup RpcBackendOptions Constructor API"


Before this commit, applications have to do the following to configure
number of threads in ProcessGroup RPC backend:

```
op = ProcessGroupRpcBackendOptions()
op.rpc_timeout = rpc_timeout
op.init_method = init_method
op.num_send_recv_threads = 32
init_rpc(...., rpc_backend_options=op)
```

After this commit, it can be simplified to:

```
init_rpc(...., rpc_backend_options=ProcessGroupRpcBackendOptions(num_send_recv_threads=32))
```

Fixes #34075

Differential Revision: [D20227344](https://our.internmc.facebook.com/intern/diff/D20227344)

[ghstack-poisoned]
